x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,215 @@
+                    "PrivateKey entries where the associated certificate is a CA certificate");
+        } else {
+            final CheckedSupplier<String, Exception> createApiKeyRequestBodySupplier = () -> {
+                XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+                xContentBuilder.startObject()
+                    .field("name", "enrollment_token_API_key_" + UUIDs.base64UUID())
+                    .field("expiration", ENROLL_API_KEY_EXPIRATION)
+                    .startObject("role_descriptors")
+                    .startObject("create_enrollment_token")
+                    .field("cluster", "[" + ClusterPrivilegeResolver.ENROLL_NODE.name() + "]")
+                    .endObject()
+                    .endObject()
+                    .endObject();
+                return Strings.toString(xContentBuilder);
+            };
+
+            final CheckedSupplier<String, Exception> getHttpInfoRequestBodySupplier = () -> {
+                XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+                xContentBuilder.startObject()
+                    .endObject();
+                return Strings.toString(xContentBuilder);
+            };
+
+            try {
+                final URL url = new URL(client.getDefaultURL());
+                final HttpResponse httpResponseApiKey = client.execute("POST", createAPIKeyURL(url), user, password,
+                    createApiKeyRequestBodySupplier, is -> responseBuilder(is, terminal));
+                int httpCode = httpResponseApiKey.getHttpStatus();
+
+                if (httpCode != HttpURLConnection.HTTP_OK) {
+                    throw new UserException(ExitCodes.CANT_CREATE,
+                        "Unexpected response code [" + httpCode + "] from calling POST " + createAPIKeyURL(url));
+                }
+
+                final String apiKey = Objects.toString(httpResponseApiKey.getResponseBody().get("api_key"), "");
+                if (apiKey == null || apiKey.isEmpty()) {
+                    throw new UserException(ExitCodes.CANT_CREATE,
+                        "Could not create an api key.");
+                }
+
+                final HttpResponse httpResponseHttp = client.execute("GET", getHttpInfoURL(url), user, password,
+                    getHttpInfoRequestBodySupplier, is -> responseBuilder(is, terminal));
+                httpCode = httpResponseHttp.getHttpStatus();
+
+                if (httpCode != HttpURLConnection.HTTP_OK) {
+                    throw new UserException(ExitCodes.CANT_CREATE,
+                        "Unexpected response code [" + httpCode + "] from calling GET " + getHttpInfoURL(url));
+                }
+
+                final String address = getPublishedAddress(httpResponseHttp.getResponseBody());
+                if (address == null || address.isEmpty()) {
+                    throw new UserException(ExitCodes.CANT_CREATE,
+                        "Could not retrieve the ip address.");
```

Comment:
Can we add more meaningful error messages, either here or make sure that the code that wraps these exceptions does that ? As a user I'm thinking: What is the ip address ? whose ip address ? how did we try to retrieve it and from where ? how and why it failed ?